hpe/dl380/exporter.go

```diff
@@ -0,0 +1,553 @@
+		retryCount := i
+		if retryCount > 0 {
+			log.Error("api call "+r.URL.String()+" failed, retry #"+strconv.Itoa(retryCount), zap.Any("trace_id", ctx.Value("traceID")))
+		}
+	}
+
+	// Check that the target passed in has http:// or https:// prefixed
+	fqdn, err := url.ParseRequestURI(target)
+	if err != nil {
+		fqdn = &url.URL{
+			Scheme: config.GetConfig().OOBScheme,
+			Host:   target,
+		}
+	}
+
+	// vars for drive parsing
+	var (
+		initialURL        = "/Systems/1/SmartStorage/ArrayControllers"
+		url               = initialURL
+		chassis_url       = "/Chassis/1"
+		logicalDriveURLs  []string
+		physicalDriveURLs []string
+		nvmeDriveURLs     []string
+	)
+
+	// PARSING DRIVE ENDPOINTS
+	// Get initial JSON return of /redfish/v1/Systems/1/SmartStorage/ArrayControllers/ set to output
+	output, err := getDriveEndpoint(fqdn.String()+uri+url, target, retryClient)
+
+	// Loop through Members to get ArrayController URLs
+	if err != nil {
+		log.Error("api call "+fqdn.String()+uri+url+" failed.", zap.Any("trace_id", ctx.Value("traceID")))
+		return nil
+	}
+
+	if output.MembersCount > 0 {
+		for _, member := range output.Members {
+			// for each ArrayController URL, get the JSON object
+			newOutput, err := getDriveEndpoint(fqdn.String()+member.URL, target, retryClient)
+			if err != nil {
+				log.Error("api call "+fqdn.String()+member.URL+" failed.", zap.Any("trace_id", ctx.Value("traceID")))
+				continue
+			}
+
+			// If LogicalDrives is present, parse logical drive endpoint until all urls are found
+			if newOutput.Links.LogicalDrives.URL != "" {
+				logicalDriveOutput, err := getDriveEndpoint(fqdn.String()+newOutput.Links.LogicalDrives.URL, target, retryClient)
+				if err != nil {
+					log.Error("api call "+fqdn.String()+newOutput.Links.LogicalDrives.URL+" failed.", zap.Any("trace_id", ctx.Value("traceID")))
+					continue
+				}
+
+				if logicalDriveOutput.MembersCount > 0 {
+					// loop through each Member in the "LogicalDrive" field
+					for _, member := range logicalDriveOutput.Members {
+						// append each URL in the Members array to the logicalDriveURLs array.
+						logicalDriveURLs = append(logicalDriveURLs, member.URL)
+					}
+				}
+			}
+
+			// If PhysicalDrives is present, parse physical drive endpoint until all urls are found
+			if newOutput.Links.PhysicalDrives.URL != "" {
+				physicalDriveOutput, err := getDriveEndpoint(fqdn.String()+newOutput.Links.PhysicalDrives.URL, target, retryClient)
+
+				if err != nil {
+					log.Error("api call "+fqdn.String()+newOutput.Links.PhysicalDrives.URL+" failed.", zap.Any("trace_id", ctx.Value("traceID")))
+					continue
+				}
+				if physicalDriveOutput.MembersCount > 0 {
+					for _, member := range physicalDriveOutput.Members {
+						physicalDriveURLs = append(physicalDriveURLs, member.URL)
+					}
+				}
+			}
+		}
+	}
+
+	// parse to find NVME drives
+	chassis_output, err := getDriveEndpoint(fqdn.String()+uri+chassis_url, target, retryClient)
```

Comment:
I know I'm nitpicking but go variable names should be camel case instead of snake case

Reply:
Done!